Content.Server/Forensics/Systems/ForensicScannerSystem.cs

```diff
@@ -0,0 +1,176 @@
+        [Dependency] private readonly UserInterfaceSystem _uiSystem = default!;
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+        [Dependency] private readonly PaperSystem _paperSystem = default!;
+        [Dependency] private readonly SharedHandsSystem _handsSystem = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<ForensicScannerComponent, AfterInteractEvent>(OnAfterInteract);
+            SubscribeLocalEvent<ForensicScannerComponent, AfterInteractUsingEvent>(OnAfterInteractUsing);
+            SubscribeLocalEvent<ForensicScannerComponent, ForensicScannerPrintMessage>(OnPrint);
+            SubscribeLocalEvent<TargetScanSuccessfulEvent>(OnTargetScanSuccessful);
+            SubscribeLocalEvent<ScanCancelledEvent>(OnScanCancelled);
+        }
+
+        private void OnScanCancelled(ScanCancelledEvent ev)
+        {
+            if (ev.Component == null)
+                return;
+            ev.Component.CancelToken = null;
+        }
+
+        private void OnTargetScanSuccessful(TargetScanSuccessfulEvent ev)
+        {
+            ev.Component.CancelToken = null;
+
+            if (!TryComp<ForensicsComponent>(ev.Target, out var forensics))
+              return;
+
+            ev.Component.Fingerprints = forensics.Fingerprints.ToList();
+            ev.Component.Fibers = forensics.Fibers.ToList();
+            OpenUserInterface(ev.User, ev.Component);
+        }
+
+        private void OnAfterInteract(EntityUid uid, ForensicScannerComponent component, AfterInteractEvent args)
+        {
+            if (component.CancelToken != null)
+            {
+                component.CancelToken.Cancel();
+                component.CancelToken = null;
+            }
+
+            if (args.Target == null || !args.CanReach)
+                return;
+
+            component.CancelToken = new CancellationTokenSource();
+            _doAfterSystem.DoAfter(new DoAfterEventArgs(args.User, component.ScanDelay, component.CancelToken.Token, target: args.Target)
+            {
+                BroadcastFinishedEvent = new TargetScanSuccessfulEvent(args.User, args.Target, component),
+                BroadcastCancelledEvent = new ScanCancelledEvent(uid, component),
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                BreakOnStun = true,
+                NeedHand = true
+            });
+        }
+
+        private void OnAfterInteractUsing(EntityUid uid, ForensicScannerComponent component, AfterInteractUsingEvent args)
+        {
+            if (args.Handled || !args.CanReach)
+                return;
+
+            if (!TryComp<ForensicPadComponent>(args.Used, out var pad))
+                return;
+
+            foreach (var fiber in component.Fibers)
+            {
+                if (fiber == pad.Sample)
+                {
```

Comment:
This is kinda eh, how many fibers would a scanner expect? If it's more than like 10 it should be a hashset.

Reply:
Less than 10.